### PR TITLE
Dev loll

### PR DESCRIPTION
**Pull request by lollcat**
- add option to adjust weight in prioritised replay without additional evaluation of flow
- add option to specify number of flow forward passes when defining how long to train
- Some numerical stability improvements (to metropolis hastings transition operator, and prevent gradient steps when the gradient norm is NaN)

Further note: I found gradient norm clipping to be very important for training with AIS targetting p^2/q, as sometimes this finds points with VERY low probability under the flow, which can cause very large gradient norms (e.g. 1e6). The gradient norm should not be set too low or too high. For the GMM I found a gradient norm of 100 works well. To find this number one can run a few iterations and eyeball the gradient norm, and then pick a value a bit larger than the typical grad norm, as we mainly just want to prevent iterations with very large gradient norms causing instability, while leaving most iterations untouched by the clipping.

### CODE REVIEW

File: examples/config/gmm.yaml [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+    n_inner_steps: 1
+  n_intermediate_distributions: 1
+
+
+training:
+  seed: 0
+  lr: 1e-3
+  batch_size: 128
+  n_iterations: null # either n_itertations or n_flow_forward_pass must have a value
+  n_flow_forward_pass: 5_000 # assumes split coupling where log prob & sample are same speed.
```
Comment (lollcat) [inline examples/config/gmm.yaml:28]: We can now specify training length by number of forward passes of the flow

File: examples/setup_run.py [hunk 2]
```diff
@@ -26,6 +26,61 @@
                            Optional[Union[ReplayBuffer, PrioritisedReplayBuffer]]], Plotter]
 
 
+def get_n_iterations(
```
Comment (lollcat) [inline examples/setup_run.py:29]: A function that gets the number of training iterations, with an option to calculate this based of the number of flow forward passes required.

File: fab/sampling_methods/transition_operators/metropolis.py [hunk 3]
```diff
@@ -47,6 +47,8 @@ def transition(self, x: torch.Tensor, log_p_x: LogProbFunc, i: int) -> torch.Ten
             acceptance_probability = torch.exp(x_proposed_log_prob - x_prev_log_prob)
             # not that sometimes this will be greater than one, corresonding to 100% probability of
             # acceptance
+            acceptance_probability = torch.nan_to_num(acceptance_probability, nan=0.0, posinf=0.0,
```
Comment (lollcat) [inline fab/sampling_methods/transition_operators/metropolis.py:50]: Prevent Nans/infs from effecting the acceptance probability in the wrong way (typically these result from too big a step size, so we treat them in a way that generally decreases the step size).

File: fab/target_distributions/gmm.py [hunk 4]
```diff
@@ -46,16 +46,24 @@ def to(self, device):
...
         return self.sample((self.n_test_set_samples, ))
 
     def log_prob(self, x: torch.Tensor):
-        return self.distribution.log_prob(x)
+        log_prob = self.distribution.log_prob(x)
+        # Very low probability samples can cause issues (we turn off validate_args of the
+        # distribution object which typically raises an expection related to this.
+        # We manually decrease the distributions log prob to prevent them having an effect on
+        # the loss/buffer.
+        log_prob[log_prob < -1e4] = -torch.tensor(float("inf"))
```
Comment (lollcat) [inline fab/target_distributions/gmm.py:65]: improve numerical stability given that we see validate_args to False for the distribution

File: fab/target_distributions/many_well.py [hunk 5]
```diff
@@ -79,6 +85,33 @@ def __init__(self, dim=4, use_gpu: bool = True, *args, **kwargs):
                 self.device = "cpu"
         else:
             self.device = "cpu"
+        x, log_w = self.create_2d_test_set_with_ais(n_intermediate_distributions,
+                                                    ais_test_set_size)
+        self.register_buffer("ais_x", x)
+        self.register_buffer("ais_log_w", log_w)
+
+
+    def create_2d_test_set_with_ais(self, n_itermediate_distributions, test_set_size):
```
Comment (lollcat) [inline fab/target_distributions/many_well.py:94]: create test set generated by AIS on the 2D many well

File: fab/train_with_prioritised_buffer.py [hunk 6]
```diff
@@ -94,31 +99,28 @@ def run(self,
...
+                w_adjust_pre_clip = torch.exp(log_w_adjust)  # no grad
                 w_adjust = torch.clip(w_adjust_pre_clip, max=self.max_adjust_w_clip)
                 # manually calculate the new form of the loss
                 loss = - torch.mean(w_adjust * log_q_x)
                 if not torch.isnan(loss) and not torch.isinf(loss):
                     loss.backward()
                     grad_norm = torch.nn.utils.clip_grad_norm_(self.model.parameters(),
                                                                self.max_gradient_norm)
-                    self.optimizer.step()
+                    if torch.isfinite(grad_norm):
```
Comment (lollcat) [inline fab/train_with_prioritised_buffer.py:111]: Improve stability

File: fab/train_with_prioritised_buffer.py [hunk 7]
```diff
@@ -94,31 +99,28 @@ def run(self,
...
-                        self.flow_device)
-                    log_q_new = self.model.flow.log_prob(x)
-                    log_adjust_insert = log_q_old - log_q_new
-                    self.buffer.adjust(log_adjust_insert, log_q_new, indices)
 
-            # we log info from the step of the recently generated ais points.
-            info = self.model.get_iter_info()
+                # Adjust log weights in the buffer.
+                if not self.w_adjust_in_buffer_after_update:
+                    with torch.no_grad():
```
Comment (lollcat) [inline fab/train_with_prioritised_buffer.py:120]: This is the default now, so adjustment does not cost us extra compute

File: fab/train_with_prioritised_buffer.py [hunk 8]
```diff
@@ -127,9 +129,24 @@ def run(self,
                         w_adjust_mean=torch.mean(w_adjust_pre_clip).detach().cpu().item(),
                         w_adjust_min=torch.min(w_adjust_pre_clip).detach().cpu().item(),
                         w_adjust_max=torch.max(w_adjust_pre_clip).detach().cpu().item(),
-                        log_w_adjust_insert_mean=torch.mean(log_adjust_insert).detach().cpu().item(),
-                        log_q_mean=torch.mean(log_q_new).detach().cpu().item()
+                        log_q_x_mean=torch.mean(log_q_x).cpu().item()
                         )
+
+            if self.w_adjust_in_buffer_after_update:
```
Comment (lollcat) [inline fab/train_with_prioritised_buffer.py:135]: False by default, won't get used

File: fab/utils/prioritised_replay_buffer.py [hunk 9]
```diff
@@ -43,14 +45,15 @@ def __init__(self, dim: int,
         self.current_index = 0
         self.is_full = False  # whether the buffer is full
         self.can_sample = False  # whether the buffer is full enough to begin sampling
+        self.sample_with_replacement = sample_with_replacement
```
Comment (lollcat) [inline fab/utils/prioritised_replay_buffer.py:48]: Set to false by default (as we don't want to have repeated samples in the batch passed to the flow update)

Comment (lollcat) [reply fab/utils/prioritised_replay_buffer.py:48]: I have added the option now if we want to play around, but is default set to what we had before

File: examples/setup_run.py (hunk 2 above)
Comment (VincentStimper) [reply examples/setup_run.py:29]: Don't we need more log prob evaluations when using HMC? For each iteration, we need `n_transition_operator_inner_steps + 1` for the leapfrog and then an additional one to do decide about acceptance. For the latter, we also need an additional step in the beginning, so it should be `(n_transition_operator_inner_steps + 1) * n_intermediate_ais_dist + 1` evaluations, right? And what about sampling?

Comment (lollcat) [reply examples/setup_run.py:29]: Yeah - good spot!

Comment (lollcat) [reply examples/setup_run.py:29]: So for HMC it should be

Comment (lollcat) [reply examples/setup_run.py:29]: Yeah so for both hmc and metropolis we need an additional forward pass for the initial sampling from the flow (via `sample_and_log_prob` - I've updated the code to fix this. But for HMC we do not need `n_transition_operator_inner_steps + 1` for the following reason: For each x point we could call the equivalent of jax's `value_and_grad` method when computing the forward pass of the flow (either for sampling to init the AIS chain, or for log prob evaluation), and then pass on the value/grad to any future parts of the code that either need grad (e.g. HMC inner step) or log prob (e.g. accept/reject criterion). Thus we just need `(n_transition_operator_inner_steps) * n_intermediate_ais_dist + 1` flow forward passes.

Comment (lollcat) [reply examples/setup_run.py:29]: I've also accounted for the batch size now

Comment (lollcat) [reply examples/setup_run.py:29]: image

Comment (lollcat) [reply examples/setup_run.py:29]: The point here is that per unique x that we generate during AIS (including part of the inner HMC step) in theory we only need one forward pass of the flow. In practice we re-do the flow forward pass many more times than what is necessary (e.g. when calculating AIS weight we typically re-evaluate the flow rather than taking the log prob that we used already in the metropolis accept/reject step).

Comment (VincentStimper) [reply examples/setup_run.py:29]: Ah, I see, you're right. Thanks for having another close look on this, @lollcat!

**Review by VincentStimper — APPROVED** [review]
Looks good now, thanks!
